### PR TITLE
Add object tags for article data references.

### PR DESCRIPTION
**Pull request by gnott**
Re issue https://github.com/elifesciences/elife-pubmed-feed/issues/78
Re issue https://github.com/elifesciences/elife-pubmed-xml-generation/issues/33

I think the changes in this PR satisfy the feature request described in the issues. After some discussion, and some confusion on my part, I understand what we want to do is: 

if there is a citation of type `"data"` that has a `<source>` tag value that matches a particular value, which we can directly match to a value PubMed will accept, then we can add it to the PubMed deposit as an `<Object>` tag.

We can continue, for now, to include the older logic that is based on `uri` value string hint matching, for datasets that are from the datasets section of the article XML.

Duplicate values are removed, based on comparing the `str` representation of the `OrderedDict` data for each, so we would only remove it if the `assigning_authority` and `params` for the dataset are exactly the same. The comparison loop is of the order O^2 I think, but the list of datasets is just a handful and should hopefull be safe for speed and resources to de-dupe.

### CODE REVIEW

File: elifepubmed/generate.py [hunk 1]
```diff
@@ -646,14 +664,44 @@ def dataset_details(dataset):
...
-            params = OrderedDict()
-            params["id"] = id_value
-            set_object(parent, assigning_authority, params)
+            dataset = OrderedDict([
+                ('assigning_authority', assigning_authority),
+                ('params', {'id': id_value})
+            ])
+            dataset_objects.append(dataset)
+    # next add from ref list but do not add duplicates
+    for ref in [ref for ref in poa_article.ref_list if ref.publication_type == 'data']:
```
Comment (lsh-0) [inline elifepubmed/generate.py:696]: this is a little crazy but surprisingly readable.

just an aside

File: tests/test_generate.py [hunk 2]
```diff
@@ -214,5 +214,34 @@ def test_set_plain_language_summary(self):
...
+            b'<root>'
+            b'<Object Type="Dryad"><Param Name="id">10.5061/dryad.example</Param></Object>'
+            b'<Object Type="NCBI:geo"><Param Name="id">GSE48760</Param></Object>'
+            b'</root>')
+        article = Article()
+        dataset = Dataset()
+        dataset.uri = '10.5061/dryad.example'
+        dataset.doi = '10.5061/dryad.example'
+        article.datasets.append(dataset)
+        # first ref is a duplicate should be ignored
```
Comment (lsh-0) [inline tests/test_generate.py:233]: is this is what is being tested? if so, stick it in the method's docstring.

**Review by lsh-0 — APPROVED** [review]

File: tests/test_generate.py (hunk 2 above)
Comment (gnott) [reply tests/test_generate.py:233]: I set out to test a bunch of values and then settled on testing both dataset and ref data with one duplicate. It is one test case to capture any edge cases that do not exist in the `00666` article XML. I'll describe it better, and I'm going to expand it to demonstrate the (soon to be) case-insensitive matching of ref source values.

**Comment by gnott** [comment]
I added in case-insensitive matching of ref source name to the source name map after one final bit if feedback from Production. Thanks for reviewing @lsh-0, since this most recent change is fairly small, I think this will be ok to merge.
